yarn-project/pxe/src/storage/private_event_data_provider/private_event_data_provider.ts

```diff
@@ -0,0 +1,137 @@
+import { sha256 } from '@aztec/foundation/crypto';
+import { Fr } from '@aztec/foundation/fields';
+import { createLogger } from '@aztec/foundation/log';
+import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
+import type { AztecAsyncArray, AztecAsyncKVStore, AztecAsyncMap } from '@aztec/kv-store';
+import type { EventSelector } from '@aztec/stdlib/abi';
+import type { AztecAddress } from '@aztec/stdlib/aztec-address';
+
+import type { DataProvider } from '../data_provider.js';
+
+interface PrivateEventEntry {
+  logContent: Buffer;
+  blockNumber: number;
+}
+
+/**
+ * Stores decrypted private event logs.
+ */
+export class PrivateEventDataProvider implements DataProvider {
+  #store: AztecAsyncKVStore;
+  /** Array storing the actual private event log entries containing the log content and block number */
+  #eventLogs: AztecAsyncArray<PrivateEventEntry>;
+  /** Map from contract_address_recipient_eventSelector to array of indices into #eventLogs for efficient lookup */
+  #eventLogIndex: AztecAsyncMap<string, number[]>;
```

Comment:
Why do it this way? I don't think we do it like this in any other store, do we?

Reply:
AI spitted all of this out and it seems fine. There was no even a consensus if we will want to have this so I gave priority to speed here.